ownCloud/Client/Actions/Actions+Extensions/DeleteAction.swift

```diff
@@ -28,8 +28,22 @@ class DeleteAction : Action {
 
 	// MARK: - Extension matching
 	override class func applicablePosition(forContext: ActionContext) -> ActionPosition {
-		guard forContext.items.sharedWithUser.count == 0 else {
-			return .none
+		let sharedWithUser = forContext.items.sharedWithUser
+
+		if let core = forContext.core {
+			if sharedWithUser.count == 1 {
+				if sharedWithUser[0].isShareRootItem(from: core) {
+					return .none
+				}
+			} else {
+				let sharedFolders = sharedWithUser.filter({$0.type == .collection})
```

Comment:
This would only check folder items. The shared root item could also be a file, though, in which case, `Delete` should not be available, either.

I see no potential to speed up that check by altering or simplifying the check itself. However, for items in the same folder, what should bring a big boost is caching test results. I've already implemented this in `item.isShareRootItem(from:)` for the `item` that the method is invoked on.

Where I still see good ways for optimization through caching here, however, is the check for the parent folder. In the `isShareRootItem(from:)` method, there's the line

```
if let parentPath = self.path?.parentPath, parentPath != self.path, let parentItem = self.parentItem(from: core) { …
```

with `self.parentItem(from: core)` triggering a queue-scheduled SQLite query, causing the slowness. For multiple items in the same folder, the parent item would always be the same. So, a possible optimization here could be achieved by:
- adding a `[OCLocalID : OCItem]` property `cachedParentFolders` to `ActionContext`.
- add a parameter that takes an `ActionContext` to `isShareRootItem()`, like `isShareRootItem(from: OCItem, in context: ActionContext? = nil)`
- in `isShareRootItem()`, use the `OCItem.parentLocalID` to check if `ActionContext.cachedParentFolders` already has an `OCItem` for the parent folder:
    - if it does, re-use that item as `parentItem`
    - if it does not, request it from the database as before, via `self.parentItem(from: core)`, then store it in `ActionContext.cachedParentFolders[parentItem.localID]` for later retrieval

Now that I've written the above, I realize that directly manipulating `ActionContext.cachedParentFolders` isn't the best/cleanest way to write this probably. A method for retrieving parent items for a particular item - like `ActionContext.parentItem(for: OCItem)` - that transparently handles caching in a dict in a private instance var would be a lot better approach IMO.

Since `ActionContext`s are short-lived, memory should not be a concern. In most cases, that cache would have only a single entry anyway.

And by caching this at the `ActionContext` level, the caching would even benefit checks from multiple actions.

Reply:
@felix-schwarz thanks for ideas.. I will revisit the implementation with them in mind..